keps/sig-autoscaling/draft-20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,170 @@
+Additionally PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not enough to handle it,
+* load has decreased significantly and currently allocated resources are unused
+  and thus wasted,
+* Resources have simply been set improperly.
+
+Currently changing Resources allocation requires the Pod to be recreated since
+the PodSpec is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating a Pod nor restarting a
+Container addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource update is not available.
+* Secondary: allow users to specify which Pods and Containers can be updated
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed an in-place resource update. These cases should be handled by
+actors which initiated the update.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to resources and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodStatus is extended with information about actually allocated resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration,
+  denoting **desired** state of the Pod,
+* PodStatus.ContainerStatus.ResourceAllocated (new object) denotes **actual**
+  state of the Pod resources.
+
+To distinguish between possible states of the Pod resources,
+a new PodCondition InPlaceResize is added, with the following states:
+* (empty) - the default value; resource update awaits reconciliation
+  (if ResourceRequirements differs from ResourceAllocated),
+* Awaiting - awaiting resources to be freed (e.g. via pre-emption),
+* Failed - resource update could not have been performed in-place
```

Comment:
What's the difference between `Failed` and `Awaiting`? Seems like they are both supposed to mean "will keep trying but not done yet."

Reply:
Yes they are very similar.

@kgolab What's the motivation to have these two distinct states? Can this be collapsed into one?

In our approach, when we set PodCondition.Status = ConditionFalse, we set PodCondition.Reason to reflect the cause such as reschedule not allowed by policy (i.e: InPlaceOnly), or PDB violation. This allows us to perform smart retries - if blocked by policy, retry when pods depart node, and if due to PDB violation, retry (currently on timer) based on PDB update. The retry approach fits well with k8s way of attempting to get actual state == desired state. We initiate resize requests and retries for failures from the controller that owns the pod, but retry can be disabled by policy if you prefer VPA to handle it.

Reply:
The distinction between Awaiting and Failed is to denote that in the Awaiting case some actions are in progress to satisfy the in place update request - e.g. some low-priority pod is being kicked out of the machine and we should wait for it to finish. This is important for VPA Updater to know that it shouldn't evict the pod just yet since there is a chance of the in place request being satisfied soon. In Failed case the request cannot be satisfied soon (e.g. no low-priority to kick out) and VPA Updater can freely evict the pod to cause reschedule to a different machine.

Reply:
To build on top of what @schylek wrote:
* Awaiting was designed specifically for pre-emption case but could be generalized to "the cluster is already taking some action to make this update possible",
* Failed was designed specifically for Kubelet to denote that in the current cluster state the update is not possible and it cannot start any action that would change this (though maybe Scheduler can).
* Rejected means that the update is not possible and we don't expect any controllers to automatically change that - for example the new resources go over Quota. This is a permanent state from the cluster point of view. If something in the cluster changes (e.g. some non-evictable Pod finishes or is rescheduled), the update will not be retried automatically.

Reply:
The updated design removes this PodCondition, so this issue is not relevant anymore. Resolving this conversation, please unresolve if any concerns.